File: admin/src/main/java/com/project/admin/exception/GlobalExceptionHandler.java
package com.project.admin.exception;

import java.util.concurrent.CompletionException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.Map;
import java.util.stream.Collectors;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(BaseException.class)
    public ResponseEntity<ErrorResponse> handleAdminException(BaseException e) {
        return ResponseEntity
                .status(e.getErrorCode().status())
                .body(ErrorResponse.from(e.getErrorCode()));
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidException(MethodArgumentNotValidException e) {

        Map<String, String> errors = e.getBindingResult().getFieldErrors().stream()
                .collect(Collectors.toMap(
                        FieldError::getField,
                        error -> error.getDefaultMessage() != null ? error.getDefaultMessage() : "유효하지 않은 값입니다",
                        (existing, replacement) -> existing
                ));

        ErrorResponse errorResponse = ErrorResponse.of(HttpStatus.BAD_REQUEST.value(), "입력값이 올바르지 않습니다", errors);


        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(errorResponse);
    }

    @ExceptionHandler(CompletionException.class)
    public ResponseEntity<ErrorResponse> handleExecutionException(CompletionException e) {

        Throwable cause = e.getCause();
        if (cause instanceof BaseException baseException) {
            return ResponseEntity
                    .status(baseException.getErrorCode().status())
                    .body(ErrorResponse.from(baseException.getErrorCode()));
        }
        ErrorResponse errorResponse = ErrorResponse.of(
                HttpStatus.INTERNAL_SERVER_ERROR.value(), "요약 실패", null);

        return ResponseEntity
                .internalServerError()
                .body(errorResponse);
    }
}

File: api/src/main/kotlin/com/project/api/controller/GlobalExceptionHandler.kt
package com.project.api.controller

import com.project.api.exception.ErrorResponse
import com.project.api.exception.CommonErrorCode
import com.project.api.exception.QuickStackException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {

    companion object {
        private val log = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)
    }

    @ExceptionHandler(QuickStackException::class)
    fun handleBusinessException(e: QuickStackException): ResponseEntity<ErrorResponse> {
        log.warn("BusinessException: {}", e.print())
        return ResponseEntity
            .status(e.errorCode.httpStatus)
            .body(ErrorResponse.from(e.errorCode))
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(e: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        e.bindingResult.fieldErrors.forEach {
            log.warn(
                "Validation failed - field: {}, value: {}, message: {}",
                it.field,
                it.rejectedValue,
                it.defaultMessage
            )
        }
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(ErrorResponse.from(CommonErrorCode.INVALID_REQUEST))
    }

    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ErrorResponse> {
        log.error("Unhandled exception", e)
        return ResponseEntity
            .status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(ErrorResponse.from(CommonErrorCode.INTERNAL_SERVER_ERROR))
    }
}